app/models/assignment/rotation_generator.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+class Assignment < ApplicationRecord
+  class RotationGenerator
+    include ActiveModel::Model
+    include ActiveModel::Attributes
+
+    attribute :roster_id, :integer
+    attribute :user_ids
+    attribute :starting_user_id, :integer
+    attribute :start_date, :date
+    attribute :end_date, :date
+
+    validates :roster, presence: true
+    validates :user_ids, presence: true
+    validates :starting_user_id, presence: true
+    validates :start_date, presence: true
+    validates :end_date, presence: true, comparison: { greater_than_or_equal_to: :start_date,
+                                                       message: :end_date_must_be_after_start }
```

Comment:
```suggestion
    validates :end_date, presence: true, comparison: { greater_than_or_equal_to: :start_date,
                                                       if: -> { start_date.present? && end_date.present? },
                                                       message: :end_date_must_be_after_start }
```
Totally missed this - we'll get redundant/weird messages if we don't ignore cases where either/both are missing